lib/assent/strategies/apple.ex

```diff
@@ -0,0 +1,99 @@
+defmodule Assent.Strategy.Apple do
+  @moduledoc """
+  Apple Sign In OAuth 2.0 strategy.
+
+  You'll need to collect the 10-char long Team ID, the Services ID, the 10-char
+  Key ID and download the private key from the portal. Save the private key to
+  an accessible folder, or alternatively set `:private_key` with the content of
+  the private key.
+
+  ## Usage
+
+      config = [
+        client_id: "REPLACE_WITH_SERVICES_ID",
+        team_id: "REPLACE_WITH_TEAM_ID",
+        private_key_id: "REPLACE_WITH_PRIVATE_KEY_ID",
+        private_key_path: "/path/to/key.p8"
+      ]
+
+  ## With JS SDK
+
+  You can use the JS SDK instead of handling it through `auhorize_url/2`. All you
+  have to do is to set up you HTML page with the following way:
+
+      <html>
+        <head>
+            <meta name="appleid-signin-client-id" content="[CLIENT_ID]">
+            <meta name="appleid-signin-scope" content="[SCOPES]">
+            <meta name="appleid-signin-redirect-uri" content="[REDIRECT_URI]">
+            <meta name="appleid-signin-state" content="[STATE]">
+        </head>
+        <body>
+            <div id="appleid-signin" data-color="black" data-border="true" data-type="sign in"></div>
+            <script type="text/javascript" src="https://appleid.cdn-apple.com/appleauth/static/jsapi/appleid/1/en_US/appleid.auth.js"></script>
+        </body>
+      </html>
+
+  You can generate a state by calling:
+
+      {:ok, %{session_params: %{state: state}}} = Assent.Strategy.Apple.authorize_url(config)
+  """
+  use Assent.Strategy.OAuth2.Base
+
+  alias Assent.{Config, Strategy.OAuth2, Strategy.OAuth2.Base}
+
+  @spec default_config(Config.t()) :: Config.t()
+  def default_config(_config) do
+    [
+      site: "https://appleid.apple.com",
+      authorize_url: "/auth/authorize",
+      token_url: "/auth/token",
+      authorization_params: [scope: "email", response_mode: "form_post"],
+      auth_method: :client_secret_post
+    ]
+  end
+
+  @spec callback(Keyword.t(), map()) :: {:ok, %{user: map()}} | {:error, term()}
+  def callback(config, params) do
+    with {:ok, client_secret} <- gen_client_secret(config) do
+      config
+      |> Config.put(:client_secret, client_secret)
+      |> Base.callback(params, __MODULE__)
+    end
+  end
+
+  def gen_client_secret(config) do
+    config =
+      config
+      |> default_config()
+      |> Keyword.merge(config)
+
+    with {:ok, site}            <- Config.fetch(config, :site),
+         {:ok, client_id}       <- Config.fetch(config, :client_id),
+         {:ok, team_id}         <- Config.fetch(config, :team_id),
+         {:ok, _private_key_id} <- Config.fetch(config, :private_key_id),
+         {:ok, private_key}     <- OAuth2.__load_private_key__(config) do
```

Comment:
What is the point of the private key id? I am confused that we don't do anything with it

Reply:
Yeah, let me see if I can rewrite this. It's automatically included in the JWT signing process if it's available, so this is just to ensure that it has been defined as it's required for Apple Sign In. It's hard to read from this.